service/controller/internal/nodecontroller/controller.go

```diff
@@ -1,108 +1,230 @@
+		return nil, microerror.Maskf(invalidConfigError, "%T.Name must not be empty", config)
+	}
+	if config.Selector == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Selector must not be empty", config)
+	}
+	if config.WorkloadK8sClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.WorkloadK8sClient must not be empty", config)
 	}
 
 	c := &Controller{
-		Controller: operatorkitController,
+		managementK8sClient: config.ManagementK8sClient,
+		workloadK8sClient:   config.WorkloadK8sClient,
+		logger:              config.Logger,
+
+		stopped:        make(chan struct{}),
+		lastReconciled: time.Time{},
+		name:           config.Name,
+		selector:       config.Selector,
+		cluster:        config.Cluster,
 	}
 
 	return c, nil
 }
 
-func newResources(config Config) ([]resource.Interface, error) {
-	var podConditionResource resource.Interface
+func (c *Controller) Boot() error {
+	var mgr manager.Manager
+
 	{
-		c := podcondition.Config{
-			Cluster:             config.Cluster,
-			ManagementK8sClient: config.ManagementK8sClient,
-			Logger:              config.Logger,
+		o := manager.Options{
+			// MetricsBindAddress is set to 0 in order to disable it. We do this
+			// ourselves.
+			MetricsBindAddress: DisableMetricsServing,
+			SyncPeriod:         to.DurationP(ResyncPeriod),
 		}
 
 		var err error
-		podConditionResource, err = podcondition.New(c)
+		mgr, err = manager.New(c.workloadK8sClient.RESTConfig(), o)
 		if err != nil {
-			return nil, microerror.Mask(err)
+			return microerror.Mask(err)
 		}
 	}
 
-	resources := []resource.Interface{
-		podConditionResource,
-	}
-
 	{
-		c := retryresource.WrapConfig{
-			Logger: config.Logger,
+		// We build our controller and set up its reconciliation.
+		// We use the Complete() method instead of Build() because we don't
+		// need the controller instance.
+		err := builder.
+			ControllerManagedBy(mgr).
+			For(&corev1.Node{}).
+			WithOptions(controller.Options{
+				MaxConcurrentReconciles: 1,
+				Reconciler:              c,
+			}).
+			WithEventFilter(predicate.Funcs{
+				CreateFunc:  func(e event.CreateEvent) bool { return c.selector.Matches(labels.Set(e.Meta.GetLabels())) },
+				DeleteFunc:  func(e event.DeleteEvent) bool { return c.selector.Matches(labels.Set(e.Meta.GetLabels())) },
+				UpdateFunc:  func(e event.UpdateEvent) bool { return c.selector.Matches(labels.Set(e.MetaNew.GetLabels())) },
+				GenericFunc: func(e event.GenericEvent) bool { return c.selector.Matches(labels.Set(e.Meta.GetLabels())) },
+			}).
+			Complete(c)
+		if err != nil {
+			return microerror.Mask(err)
 		}
 
-		var err error
-		resources, err = retryresource.Wrap(resources, c)
+		// Start in goroutine so we don't block caller
+		go func() {
+			// mgr.Start() blocks the boot process until it ends gracefully or fails.
+			err = mgr.Start(c.stopped)
+			if err != nil {
+				_, _ = fmt.Fprintf(os.Stderr, "error stopping manager: %s", microerror.Mask(err).Error())
```

Comment:
I just stepped through the code that returns an error here. It basically comes down to an error being returned if the manager child goroutines (like the metrics endpoint although that's disabled) don't shut down within the configured grace period. Since this is also run in a goroutine, the only other option I can see is to panic and os.Exit in this case to make sure we don't leak goroutines. In normal cases, this should happen rarely, but in resource-constrained or other degernate cases this could happen dozens of times per hour causing excessive resource usage. I guess I'll switch to a panic and exit if that sounds reasonable.

Reply:
By the way, found a sweet Star Trek meme in controller-runtime while looking at this https://github.com/kubernetes-sigs/controller-runtime/blob/e1a725df2743147795e5dfc8275365f7ada24805/pkg/manager/internal.go#L456

Reply:
Sounds reasonable to me. We will then be able to tell that it's happening easier at least :)

Reply:
Switched to a panic here so at least we'll know if something bad happens with an operator restart.